### ISSUE TITLE
cannot enforce type shape when assigned to a conjunction

### ISSUE THREAD

**Issue by sledorze** [microsoft/TypeScript#28642]
**TypeScript Version:**  3.1.6, 3.2.0-rc, 3.3.0-dev.20181121

**Search Terms:**
conjunction assignable wrong type

**Code**

```ts
declare const a: { a: string; b: number }
declare const b: { a: string } & { b: number }
declare function f(x: { [k: string]: string }): void
f(a) // error
f(b) // NO error
```

**Expected behavior:**
Errors on both f(a) and f(b)

**Actual behavior:**
Errors only on f(a)

**Playground Link:** http://www.typescriptlang.org/play/#src=declare%20const%20a%3A%20%7B%20a%3A%20string%3B%20b%3A%20number%20%7D%0D%0Adeclare%20const%20b%3A%20%7B%20a%3A%20string%20%7D%20%26%20%7B%20b%3A%20number%20%7D%0D%0Adeclare%20function%20f(x%3A%20%7B%20%5Bk%3A%20string%5D%3A%20string%20%7D)%3A%20void%0D%0Af(a)%20%2F%2F%20error%0D%0Af(b)%20%2F%2F%20NO%20error

**Related Issues:** https://github.com/gcanti/io-ts/issues/234#issuecomment-440752593
https://github.com/gcanti/fp-ts-routing/issues/37

**Comment by sledorze** [comment]
I'm guessing this type relaxation is required in certain scenario (with named members), however, with an indexed type on string (covering both 'a' and 'b'), I would rather expect it to fail.

**Comment by jack-williams** [comment]
- `T & U` is assignable to `T`
- `{ a: string }` is assignable to `{ [k: string]: string }`

String indexers are unsound like this when properties are 'forgotten'. Another example:
- `{ a : string; b: number }` is assignable to `{ a: string }`
- `{ a: string }` is assignable to `{ [k: string]: string }`
```ts
const a: { a: string; b: number } = { a: 'foo', b: 42 };
const b: { a: string } = a
const c: { [k: string]: string } = c;
c.b = "not a number"
const shouldBeNumber: number = a.b;
```

**Comment by sledorze** [comment]
@jack-williams indeed.
The amount of consequences of the unsoundness is frightening.

**Comment by MartinJohns** [comment]
@jack-williams Quoting the spec is not advised, as the spec was last updated around **3 years** ago.

**Comment by jakebailey** [comment]
This was fixed by #35143, whose description contains effectively this issue's repro
